File: samples/button_sidebar/sidebarbutton.nim
import ../../vendor/nim_miqt_poc/seeqt/[
  qpushbutton, qicon, qsize, qwidget
]

type
  SidebarButton* = ref object
    button*: QPushButton
    active*: bool

proc newSidebarButton*(text: string, icon: QIcon, parent: QWidget): SidebarButton =
  ## Create a new sidebar button
  result = SidebarButton(
    button: QPushButton.create(text, parent),
    active: false
  )
  
  # Set default styling
  result.button.setFixedHeight(50)
  result.button.setMinimumWidth(150)
  result.button.setCheckable(true)
  result.button.setFlat(true)
  
  # Set initial style
  result.button.setStyleSheet("QPushButton { text-align: left; padding-left: 10px; color: white; font-weight: 500; border: none; } QPushButton:hover { background-color: rgba(52, 73, 94, 180); }")
  
  # Set icon if provided
  if not icon.isNull():
    result.button.setIcon(icon)
    result.button.setIconSize(QSize.create(24, 24))

proc setActive*(self: SidebarButton, active: bool) =
  ## Set the active state of the button
  if self.active == active:
    return  # No change needed
    
  self.active = active
  self.button.setChecked(active)
  
  # Update the style based on active state
  if active:
    # Active button has a different background - lighter for contrast against dark sidebar
    self.button.setStyleSheet("QPushButton { text-align: left; padding-left: 10px; color: white; font-weight: 500; border: none; background-color: rgba(70, 130, 180, 100); border-left: 5px solid #3498db; } QPushButton:hover { background-color: rgba(70, 130, 180, 120); }")
  else:
    # Normal state with hover effect
    self.button.setStyleSheet("QPushButton { text-align: left; padding-left: 10px; color: white; font-weight: 500; border: none; } QPushButton:hover { background-color: rgba(52, 73, 94, 180); }")

proc isActive*(self: SidebarButton): bool =
  ## Check if the button is active
  return self.active

proc sizeHint*(self: SidebarButton): QSize =
  ## Return the size hint for the button
  return self.button.sizeHint()

File: reference/button_sidebar/sidebarbutton.cpp
#include "sidebarbutton.h"
#include <QPainter>
#include <QPaintEvent>
#include <QStyleOption>

SidebarButton::SidebarButton(const QString &text, const QIcon &icon, QWidget *parent)
    : QPushButton(parent), m_active(false)
{
    setText(text);
    setIcon(icon);
    setIconSize(QSize(24, 24));
    setCheckable(true);
    setFlat(true);
    
    // Set fixed height for the button
    setFixedHeight(50);
    
    // Set minimum width
    setMinimumWidth(150);
    
    // Left align text with icon and make text color white for better visibility on dark background
    setStyleSheet("text-align: left; padding-left: 10px; color: white; font-weight: 500;");
}

void SidebarButton::setActive(bool active)
{
    m_active = active;
    setChecked(active);
    update();
}

bool SidebarButton::isActive() const
{
    return m_active;
}

void SidebarButton::paintEvent(QPaintEvent *event)
{
    QPainter painter(this);
    QStyleOption opt;
    opt.initFrom(this);
    
    // Draw the button background
    if (m_active) {
        // Active button has a different background - lighter for contrast against dark sidebar
        painter.fillRect(rect(), QColor(70, 130, 180, 100));  // Steel blue with more opacity
        
        // Draw a vertical indicator on the left side
        painter.fillRect(QRect(0, 0, 5, height()), QColor(52, 152, 219));  // Bright blue indicator
    } else if (underMouse()) {
        // Hover state - lighter for better visibility on dark background
        painter.fillRect(rect(), QColor(52, 73, 94, 180));  // Slightly lighter than sidebar
    }
    
    // Draw the icon and text
    QPushButton::paintEvent(event);
}

QSize SidebarButton::sizeHint() const
{
    return QSize(150, 50);
} 